File: problems/0001_two-sum/solution.rb
#!/usr/bin/ruby

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
    m = Hash.new(nums.length)
    nums.each_with_index do |num, i|
        r = target - num
        if (m.has_key?(r))
            return [m[r], i]
        end
        m[num] = i
    end
    raise 'No match value for #{nums}, #{target}!'
end

File: problems/0001_two-sum/solution.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
    'use strict'
    const m = new Map()
    for (let i = 0; i < nums.length; ++i) {
        const num = nums[i],
              j = m.get(target - num)
        if (j !== undefined)
            return [j, i]
        m.set(num, i)
    }
    throw `No match value for ${nums}, ${target}!`
};
